File: 19. Graph - I/2. Graph/Class Code/1. Undirected Graph Cycle Using BFS & DFS.cpp
#include <iostream>
#include<vector>
#include<unordered_map>
#include<list>
#include<queue>
using namespace std;
template <typename T>
class Graph {
public:
	unordered_map<T, list<T > > adjList;

	void addEdge(T u, T v, bool direction) {
		//direction = 0 -> undirected graph
		//direction = 1 -> directed graph
		//create an edge from u to v
		adjList[u].push_back(v);
		if(direction == 0)
		{
			//undirected edge
			//create an edge from v to u
			adjList[v].push_back(u);
		}
	}

	void printAdjacencyList() {
		for(auto node: adjList) {
			cout << node.first << "-> " ;
			for(auto neighbour: node.second) {
				cout <<neighbour<<", ";
			}
			cout << endl;
		}
	}

	bool checkCyclicUsingBfs(int src, unordered_map<int,bool>& visited) {
		queue<int> q;
		unordered_map<int,int> parent;
		
		q.push(src);
		visited[src] = true;
		parent[src] = -1;

		while(!q.empty()) {
			int frontNode = q.front();
			q.pop();

			for(auto nbr: adjList[frontNode]) {
				if(!visited[nbr]) {
					q.push(nbr);
					visited[nbr] = true;
					parent[nbr]=frontNode;
				}
				if(visited[nbr] && nbr != parent[frontNode]) {
						//cycle present
						return true;
				}
			}
		}
		return false;
	}

	bool checkCyclicUsingDfs(int src, unordered_map<int,bool>& visited, int parent) {
		visited[src] = true;

		for(auto nbr: adjList[src]) {
			if(!visited[nbr]) {
				bool checkAageKaAns = checkCyclicUsingDfs(nbr, visited, src);
				if(checkAageKaAns == true)
					return true;
			}
			if(visited[nbr] && nbr != parent) {
				//cycle present
				return true;
			}
		}
		return false;
	}
};


int main() {

	Graph<int> g;
	//n -> number of nodes in graph
	int n = 5;
	g.addEdge(0,1,1);
	g.addEdge(1,2,1);
	g.addEdge(2,3,1);
	g.addEdge(3,4,1);
	g.addEdge(4,0,1);

	g.printAdjacencyList();
	cout << endl;


	bool ans = false;
	unordered_map<int, bool> visited;
	for(int i=0; i<n; i++) {
		if(!visited[i]) {
			 ans = g.checkCyclicUsingBfs(i,visited);
			if(ans == true)
				break;
		}
	}

	if(ans == true) 
		cout << "Cycle is Present" << endl;
	else
		cout << "Cycle Absent" << endl;


	// bool ansDfs = false;
	// unordered_map<int, bool> visitedDfs;
	// for(int i=0; i<n; i++) {
	// 	if(!visitedDfs[i]) {
	// 		 ansDfs = g.checkCyclicUsingDfs(i,visitedDfs, -1);
	// 		if(ansDfs == true)
	// 			break;
	// 	}
	// }

	// if(ansDfs == true) 
	// 	cout << "Cycle is Present" << endl;
	// else
	// 	cout << "Cycle Absent" << endl;

	return 0;
}

File: 19. Graph - I/2. Graph/1. Undirected Graph Cycle Using BFS & DFS.py
from collections import defaultdict, deque

class Graph:
    def __init__(self):
        self.adjList = defaultdict(list)

    def addEdge(self, u, v, direction):
        # direction = False -> undirected graph
        # direction = True -> directed graph
        # create an edge from u to v
        self.adjList[u].append(v)
        if not direction:
            # undirected edge
            # create an edge from v to u
            self.adjList[v].append(u)

    def printAdjacencyList(self):
        for node in self.adjList:
            print(node, "->", end=" ")
            for neighbour in self.adjList[node]:
                print(neighbour, end=", ")
            print()

    def checkCyclicUsingBfs(self, src, visited):
        q = deque()
        parent = {}
        
        q.append(src)
        visited[src] = True
        parent[src] = -1

        while q:
            frontNode = q.popleft()

            for nbr in self.adjList[frontNode]:
                if not visited[nbr]:
                    q.append(nbr)
                    visited[nbr] = True
                    parent[nbr] = frontNode
                if visited[nbr] and nbr != parent[frontNode]:
                    # cycle present
                    return True

        return False

    def checkCyclicUsingDfs(self, src, visited, parent):
        visited[src] = True

        for nbr in self.adjList[src]:
            if not visited[nbr]:
                checkAageKaAns = self.checkCyclicUsingDfs(nbr, visited, src)
                if checkAageKaAns:
                    return True
            if visited[nbr] and nbr != parent:
                # cycle present
                return True

        return False

g = Graph()
n = 5  # number of nodes in the graph
g.addEdge(0, 1, True)
g.addEdge(1, 2, True)
g.addEdge(2, 3, True)
g.addEdge(3, 4, True)
g.addEdge(4, 0, True)

g.printAdjacencyList()
print()

# # BFS Approach
# ans = False
# visited = defaultdict(bool)
# for i in range(n):
#     if not visited[i]:
#         ans = g.checkCyclicUsingBfs(i, visited)
#         if ans:
#             break

# if ans:
#     print("Cycle is Present")
# else:
#     print("Cycle Absent")

# DFS Approach
ansDfs = False
visitedDfs = defaultdict(bool)
for i in range(n):
    if not visitedDfs[i]:
        ansDfs = g.checkCyclicUsingDfs(i, visitedDfs, -1)
        if ansDfs:
            break

if ansDfs:
    print("Cycle is Present")
else:
    print("Cycle Absent")
